ansible_galaxy/actions/publish.py

```diff
@@ -0,0 +1,89 @@
+
+import hashlib
+import json
+import logging
+import os
+import tarfile
+
+from ansible_galaxy.rest_api import GalaxyAPI
+from ansible_galaxy import collection_artifact_manifest
+from ansible_galaxy import exceptions
+
+log = logging.getLogger(__name__)
+
+
+def _get_file_checksum(file_path):
+    checksum = hashlib.sha256()
+    with open(file_path, "rb") as f:
+        for byte_block in iter(lambda: f.read(4096), b""):
+            checksum.update(byte_block)
+    return checksum.hexdigest()
+
+
+def _publish(galaxy_context,
+             archive_path,
+             display_callback=None):
+
+    results = {
+        'errors': [],
+        'success': True
+    }
+
+    archive = tarfile.open(archive_path, 'r')
+    top_dir = os.path.commonprefix(archive.getnames())
+    manifest_path = os.path.join(top_dir,
+                                 collection_artifact_manifest.COLLECTION_MANIFEST_FILENAME)
+    try:
+        manifest_file = archive.extractfile(manifest_path)
+    except Exception as exc:
+        raise exceptions.GalaxyPublishError(str(exc), archive_path=archive_path)
+
+    try:
+        manifest = collection_artifact_manifest.load(manifest_file)
+    except Exception as exc:
+        raise exceptions.GalaxyPublishError(str(exc), archive_path=archive_path)
+
+    display_callback(json.dumps(manifest.collection_info))
+
+    api = GalaxyAPI(galaxy_context)
+
+    collection_name = manifest.collection_info.get('name')
+    try:
+        namespace_name = collection_name.split('.', 1)[0]
```

Comment:
Doesn't seem like the right place to have to deal with parsing the namespace/name. CollectionInfo should provide attributes or properties for each directly and when joined. Validation on CollectionInfo could catch any bogus formats when it is created as  part of the manifest load.